File: test/alu_test.cpp
// Test file for component of Section 1.3.1: Arithmetic Logic Unit

#include <alu.h>
#include <alu_control.h>
#include <gtest/gtest.h>
#include <verilated.h>

class ALU : public ::testing::Test {
   protected:
    alu* alu_dut;

    void AdvanceClock() {
        alu_dut->clock = 1;
        alu_dut->eval();
        alu_dut->clock = 0;
        alu_dut->eval();
    }

    void SetUp() {
        alu_dut = new alu;
        alu_dut->eval();
    }

    void TearDown() {
        alu_dut->final();
        delete alu_dut;
    }
};

TEST_F(ALU, TuringRequirement2111) {
    alu_dut->register1 = 3;
    alu_dut->register2 = 2;
    alu_dut->enable = 1;
    alu_dut->op = alu_control::alu_op_e::ADD;

    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 5);

    alu_dut->register1 = 0b11111111;  // 255
    alu_dut->register2 = 0b00000001;  // 1

    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 0);

    alu_dut->register1 = 0b11111111;  // 255
    alu_dut->register2 = 0b00000010;  // 2

    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 1);
}

TEST_F(ALU, FeatureRequirement2111) {
    alu_dut->register1 = 5;
    alu_dut->register2 = 3;

    alu_dut->enable = 1;
    alu_dut->op = alu_control::alu_op_e::SUB;

    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 2);
}

TEST_F(ALU, FreatureRequirement2112) {
    alu_dut->register1 = 0b10110101;  // 181

    alu_dut->enable = 1;
    alu_dut->op = alu_control::alu_op_e::SHL;

    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 0b01101010);  // 106

    alu_dut->register1 = 0b10110101;  // 181

    alu_dut->enable = 1;
    alu_dut->op = alu_control::alu_op_e::SHR;

    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 0b01011010);  // 90
}

TEST_F(ALU, FreatureRequirement2113) {
    alu_dut->register1 = 0b10110101;  // 181

    alu_dut->enable = 1;
    alu_dut->op = alu_control::alu_op_e::ROL;

    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 0b01101011);  // 107

    alu_dut->register1 = 0b10110101;  // 181
    alu_dut->enable = 1;
    alu_dut->op = alu_control::alu_op_e::ROR;
    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 0b11011010);  // 218
}

TEST_F(ALU, FreatureRequirement2114) {
    alu_dut->register1 = 0b01010111;
    alu_dut->register2 = 0b00110010;

    alu_dut->enable = 1;
    alu_dut->op = alu_control::alu_op_e::AND;

    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 0b00010010);

    alu_dut->op = alu_control::alu_op_e::OR;

    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 0b01110111);

    alu_dut->op = alu_control::alu_op_e::XOR;

    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 0b01100101);
}

TEST_F(ALU, FreatureRequirement2115) {
    alu_dut->register1 = 0b10101010;

    alu_dut->enable = 1;

    alu_dut->op = alu_control::alu_op_e::NOT;

    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 0b01010101);
}

TEST_F(ALU, TuringRequirement2112) {
    alu_dut->register1 = 1;
    alu_dut->register2 = 1;
    alu_dut->enable = 1;
    alu_dut->op = alu_control::alu_op_e::ADD;

    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 2);
    EXPECT_EQ(alu_dut->flag, alu_control::alu_flag_e::NONE);

    alu_dut->register2 = 255;
    alu_dut->register1 = 255;

    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 254);
    EXPECT_EQ(alu_dut->flag, alu_control::alu_flag_e::CARRY);

}

TEST_F(ALU, FeatureRequirement2116) {
    alu_dut->register1 = 1;
    alu_dut->register2 = 1;
    alu_dut->enable = 1;
    alu_dut->op = alu_control::alu_op_e::SUB;

    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 0);
    EXPECT_EQ(alu_dut->flag, alu_control::alu_flag_e::ZERO);
}

TEST_F(ALU, FreatureRequirement2117) {
    alu_dut->register1 = 12;
    alu_dut->register2 = 2;

    alu_dut->enable = 1;
    alu_dut->op = 16;

    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 0);
}

TEST_F(ALU, FreatureRequirement2118) {
    alu_dut->register1 = 12;
    alu_dut->register2 = 2;

    alu_dut->enable = 0;

    alu_dut->op = alu_control::alu_op_e::ADD;

    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 0);

    alu_dut->enable = 1;

    AdvanceClock();

    alu_dut->enable = 0;

    AdvanceClock();

    EXPECT_EQ(alu_dut->result, 0);
}


int main(int argc, char** argv) {
    Verilated::commandArgs(argc, argv);
    testing::InitGoogleTest(&argc, argv);
    auto res = RUN_ALL_TESTS();
    Verilated::mkdir("logs");
    VerilatedCov::write("logs/alu.dat");
    return res;
}

File: test/alu_test.py
import cocotb
import pytest
from controlpack import ALU_OP
from utils import advance_sim
from collections import namedtuple

Param = namedtuple('Param', ["op", "reg1", "reg2", "expected_result", "expected_carry", "expected_zero", "test_name"])

@cocotb.test()
async def test_turing_req_2111_arithmetics(dut):
    params = [
        Param(ALU_OP.ADD, 0, 0, 0, 0, 1, "Zero"),  # Zero
        Param(ALU_OP.ADD, 3, 2, 5, 0, 0, "Basic functionality"),  # Basic functionality
        Param(ALU_OP.ADD, 255, 1, 0, 1, 1, "Roll over, carry flag, zero flag"),  # Roll over, carry flag, zero flag
        Param(ALU_OP.ADD, 255, 2, 1, 1, 0, "Roll over, carry flag"),  # Roll over, carry flag

        Param(ALU_OP.SUB, 0, 0, 0, 0, 1, "Zero"),  # Zero
        Param(ALU_OP.SUB, 5, 3, 2, 0, 0, "Basic functionality"),  # Basic functionality
        Param(ALU_OP.SUB, 0, 1, 255, 1, 0, "Borrow, carry flag"),  # Borrow, carry flag
        Param(ALU_OP.SUB, 1, 1, 0, 0, 1, "Result zero, zero flag")  # Result zero, zero flag
    ]
    for param in params:
        dut.register1.value = param.reg1
        dut.register2.value = param.reg2
        dut.op.value = param.op.value  # Assuming ADD is encoded as 0

        await advance_sim(dut)
        assert dut.result.value == param.expected_result, f"Failed: {param.test_name}"
        assert dut.is_zero.value == param.expected_zero, f"Failed: {param.test_name}"
        assert dut.is_carry.value == param.expected_carry, f"Failed: {param.test_name}"

@cocotb.test()
async def test_feature_req_2114_logic(dut):
    params = [
        # AND operation
        Param(ALU_OP.AND, 0b01010111, 0b00110010, 0b00010010, 0, 0, "AND Basic functionality"),  # Basic functionality
        Param(ALU_OP.AND, 0b11111111, 0b00000000, 0b00000000, 0, 1, "AND Zero result"),  # Zero result
        Param(ALU_OP.AND, 0b11111111, 0b11111111, 0b11111111, 0, 0, "AND All bits set"),  # All bits set

        # OR operation
        Param(ALU_OP.OR, 0b01010111, 0b00110010, 0b01110111, 0, 0, "OR Basic functionality"),  # Basic functionality
        Param(ALU_OP.OR, 0b00000000, 0b00000000, 0b00000000, 0, 1, "OR Zero result"),  # Zero result
        Param(ALU_OP.OR, 0b11111111, 0b00000000, 0b11111111, 0, 0, "OR One operand all bits set"),  # One operand all bits set

        # XOR operation
        Param(ALU_OP.XOR, 0b01010111, 0b00110010, 0b01100101, 0, 0, "XOR Basic functionality"),  # Basic functionality
        Param(ALU_OP.XOR, 0b00000000, 0b00000000, 0b00000000, 0, 1, "XOR Zero result"),  # Zero result
        Param(ALU_OP.XOR, 0b11111111, 0b11111111, 0b00000000, 0, 1, "XOR All bits set, zero result"),  # All bits set, zero result

        # NOT operation
        Param(ALU_OP.NOT, 0b10101010, 0, 0b01010101, 0, 0, "NOT Basic functionality"),  # Basic functionality
        Param(ALU_OP.NOT, 0b00000000, 0, 0b11111111, 0, 0, "NOT Zero input"),  # Zero input
        Param(ALU_OP.NOT, 0b11111111, 0, 0b00000000, 0, 1, "NOT All bits set, zero result"),  # All bits set, zero result
    ]

    for param in params:
        dut.register1.value = param.reg1
        dut.register2.value = param.reg2
        dut.op.value = param.op.value

        await advance_sim(dut)
        assert dut.result.value == param.expected_result, f"Failed: {param.test_name}"
        assert dut.is_zero.value == param.expected_zero, f"Failed: {param.test_name}"
        assert dut.is_carry.value == param.expected_carry, f"Failed: {param.test_name}"

@cocotb.test()
async def test_feature_req_2115_shift(dut):
    params = [
        # ROR operation
        Param(ALU_OP.ROR, 0b00000001, 0, 0b10000000, 0, 0, "ROR Rotate right"),  # Rotate right
        Param(ALU_OP.ROR, 0b10000001, 0, 0b11000000, 0, 0, "ROR Rotate right with carry"),  # Rotate right with carry
        Param(ALU_OP.ROR, 0b00000000, 0, 0b00000000, 0, 1, "ROR Zero input"),  # Zero input

        # ROL operation
        Param(ALU_OP.ROL, 0b10000000, 0, 0b00000001, 0, 0, "ROL Rotate left"),  # Rotate left
        Param(ALU_OP.ROL, 0b11000000, 0, 0b10000001, 0, 0, "ROL Rotate left with carry"),  # Rotate left with carry
        Param(ALU_OP.ROL, 0b00000000, 0, 0b00000000, 0, 1, "ROL Zero input"),  # Zero input

        # SHR operation
        Param(ALU_OP.SHR, 0b00000010, 0, 0b00000001, 0, 0, "SHR Shift right"),  # Shift right
        Param(ALU_OP.SHR, 0b00000001, 0, 0b00000000, 1, 1, "SHR Shift right to zero"),  # Shift right to zero
        Param(ALU_OP.SHR, 0b00000000, 0, 0b00000000, 0, 1, "SHR Zero input"),  # Zero input

        # SHL operation
        Param(ALU_OP.SHL, 0b00000001, 0, 0b00000010, 0, 0, "SHL Shift left"),  # Shift left
        Param(ALU_OP.SHL, 0b10000000, 0, 0b00000000, 1, 1, "SHL Shift left with carry"),  # Shift left with carry
        Param(ALU_OP.SHL, 0b00000000, 0, 0b00000000, 0, 1, "SHL Zero input"),  # Zero input
    ]

    for param in params:
        dut.register1.value = param.reg1
        dut.register2.value = param.reg2
        dut.op.value = param.op.value

        await advance_sim(dut)
        assert dut.result.value == param.expected_result, f"Failed: {param.test_name}"
        assert dut.is_zero.value == param.expected_zero, f"Failed: {param.test_name}"
        assert dut.is_carry.value == param.expected_carry, f"Failed: {param.test_name}"
